src/basic/process-util.c

```diff
@@ -2054,14 +2060,57 @@ int posix_spawn_wrapper(const char *path, char *const *argv, char *const *envp,
         r = posix_spawnattr_init(&attr);
         if (r != 0)
                 return -r; /* These functions return a positive errno on failure */
-        /* Set all signals to SIG_DFL */
-        r = posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGMASK|POSIX_SPAWN_SETSIGDEF);
+
+#if HAVE_PIDFD_SPAWN
+        if (cgroup) {
+                r = cg_get_path_and_check(
+                                SYSTEMD_CGROUP_CONTROLLER,
+                                cgroup,
+                                /* suffix= */ NULL,
+                                &resolved_cgroup);
+                if (r < 0) {
+                        r = -r;
+                        goto fail;
+                }
+
+                cgroup_fd = open(resolved_cgroup, O_PATH|O_DIRECTORY|O_CLOEXEC);
+                if (cgroup_fd < 0) {
+                        r = errno;
+                        goto fail;
+                }
+
+                r = posix_spawnattr_setcgroup_np(&attr, cgroup_fd);
+                if (r != 0)
+                        goto fail;
+
+                flags |= POSIX_SPAWN_SETCGROUP;
+        }
+#endif
+
+        r = posix_spawnattr_setflags(&attr, flags);
         if (r != 0)
                 goto fail;
         r = posix_spawnattr_setsigmask(&attr, &mask);
         if (r != 0)
                 goto fail;
 
+#if HAVE_PIDFD_SPAWN
+        r = pidfd_spawn(&pidfd, path, NULL, &attr, argv, envp);
+        if (r == 0) {
+                posix_spawnattr_destroy(&attr);
+                return pidref_set_pidfd_consume(ret_pidref, TAKE_FD(pidfd));
```

Comment:
I'd like this to return 1 if we've managed to spawn + moving to target cgroup, and only call `cg_attach` ourselves if returned 0

Reply:
Is it worth it? It would need a lot of machinery to skip the attach done in the child process

Reply:
Ah, I only mean the parent.